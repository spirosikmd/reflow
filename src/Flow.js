import React, { PureComponent } from 'react';
import { Popup } from 'semantic-ui-react';

const style = {
  connector: {
    stroke: 'rgb(0, 0, 0)',
    strokeWidth: 1,
  },
  timeline: {
    stroke: 'rgb(208, 208, 208)',
    strokeWidth: 1,
  },
};

const timelines = {
  featureOne: 50,
  featureTwo: 110,
  develop: 170,
  release: 230,
  hotfix: 290,
  master: 350,
  top: 30,
};

class Flow extends PureComponent {
  constructor(props) {
    super(props);

    this.renderUpdateMethods = {
      master: this.renderMasterUpdate.bind(this),
      develop: this.renderDevelopUpdate.bind(this),
      hotfix: this.renderHotfixUpdate.bind(this),
      release: this.renderReleaseUpdate.bind(this),
      featureOne: this.renderFeatureOneUpdate.bind(this),
      featureTwo: this.renderFeatureTwoUpdate.bind(this),
    };

    this.renderConnectorMethods = {
      master: {
        master: this.connectMasterToMaster.bind(this),
        develop: this.connectMasterToDevelop.bind(this),
      },
      develop: {
        develop: this.connectDevelopToDevelop.bind(this),
        release: this.connectDevelopToRelease.bind(this),
        featureOne: this.connectDevelopToFeatureOne.bind(this),
        featureTwo: this.connectDevelopToFeatureTwo.bind(this),
      },
      hotfix: {
        develop: this.connectHotfixToDevelop.bind(this),
        master: this.connectHotfixToMaster.bind(this),
      },
      release: {
        release: this.connectRelaseToRelease.bind(this),
        develop: this.connectReleaseToDevelop.bind(this),
        master: this.connectReleaseToMaster.bind(this),
      },
      featureTwo: {
        featureTwo: this.connectFeatureTwoToFeatureTwo.bind(this),
        develop: this.connectFeatureTwoToDevelop.bind(this),
      },
      featureOne: {
        featureOne: this.connectFeatureOneToFeatureOne.bind(this),
        develop: this.connectFeatureOneToDevelop.bind(this),
      },
    };
  }

  renderUpdate(cx, cy, fill, popup = false, content = '') {
    if (popup) {
      return (
        <Popup
          trigger={
            <circle
              cx={cx}
              cy={cy}
              r="10"
              stroke="black"
              strokeWidth="2"
              fill={fill}
            />
          }
          content={content}
          position="bottom center"
        />
      );
    }
    return <circle cx={cx} cy={cy} r="10" stroke="black" fill={fill} />;
  }

  renderDevelopUpdate(cy, popup, content) {
    return this.renderUpdate(timelines.develop, cy, '#FEE148', popup, content);
  }

  renderMasterUpdate(cy, popup, content) {
    return this.renderUpdate(timelines.master, cy, '#39E4F7', popup, content);
  }

  renderHotfixUpdate(cy, popup, content) {
    return this.renderUpdate(timelines.hotfix, cy, '#FB5B68', popup, content);
  }

  renderReleaseUpdate(cy, popup, content) {
    return this.renderUpdate(timelines.release, cy, '#57C132', popup, content);
  }

  renderFeatureOneUpdate(cy, popup, content) {
    return this.renderUpdate(
      timelines.featureOne,
      cy,
      '#FC73BF',
      popup,
      content
    );
  }

  renderFeatureTwoUpdate(cy, popup, content) {
    return this.renderUpdate(
      timelines.featureTwo,
      cy,
      '#FC73BF',
      popup,
      content
    );
  }

  renderConnector(x1, y1, x2, y2) {
    return <line x1={x1} y1={y1} x2={x2} y2={y2} style={style.connector} />;
  }

  renderTimeline(x1, y1) {
    return (
      <line x1={x1} y1={y1} x2={x1} y2={y1 + 650} style={style.timeline} />
    );
  }

  connectDevelopToDevelop(cy1, cy2) {
    return this.renderConnector(timelines.develop, cy1, timelines.develop, cy2);
  }

  connectMasterToDevelop(cy1, cy2) {
    return this.renderConnector(timelines.master, cy1, timelines.develop, cy2);
  }

  connectHotfixToDevelop(cy1, cy2) {
    return this.renderConnector(timelines.hotfix, cy1, timelines.develop, cy2);
  }

  connectHotfixToMaster(cy1, cy2) {
    return this.renderConnector(timelines.hotfix, cy1, timelines.master, cy2);
  }

  connectDevelopToRelease(cy1, cy2) {
    return this.renderConnector(timelines.develop, cy1, timelines.release, cy2);
  }

  connectRelaseToRelease(cy1, cy2) {
    return this.renderConnector(timelines.release, cy1, timelines.release, cy2);
  }

  connectReleaseToDevelop(cy1, cy2) {
    return this.renderConnector(timelines.release, cy1, timelines.develop, cy2);
  }

  connectMasterToMaster(cy1, cy2) {
    return this.renderConnector(timelines.master, cy1, timelines.master, cy2);
  }

  connectReleaseToMaster(cy1, cy2) {
    return this.renderConnector(timelines.release, cy1, timelines.master, cy2);
  }

  connectDevelopToFeatureOne(cy1, cy2) {
    return this.renderConnector(
      timelines.develop,
      cy1,
      timelines.featureOne,
      cy2
    );
  }

  connectDevelopToFeatureTwo(cy1, cy2) {
    return this.renderConnector(
      timelines.develop,
      cy1,
      timelines.featureTwo,
      cy2
    );
  }

  connectFeatureTwoToFeatureTwo(cy1, cy2) {
    return this.renderConnector(
      timelines.featureTwo,
      cy1,
      timelines.featureTwo,
      cy2
    );
  }

  connectFeatureTwoToDevelop(cy1, cy2) {
    return this.renderConnector(
      timelines.featureTwo,
      cy1,
      timelines.develop,
      cy2
    );
  }

  connectFeatureOneToFeatureOne(cy1, cy2) {
    return this.renderConnector(
      timelines.featureOne,
      cy1,
      timelines.featureOne,
      cy2
    );
  }

  connectFeatureOneToDevelop(cy1, cy2) {
    return this.renderConnector(
      timelines.featureOne,
      cy1,
      timelines.develop,
      cy2
    );
  }

  renderUpdates(updates) {
    return updates.map(update => {
      const connectors = update.connectors;
      let connectorsToRender = [];
      if (connectors) {
        connectorsToRender = connectors.map(connector => {
          const connectorMethod = this.renderConnectorMethods[update.type][
            connector.type
          ];
          return connectorMethod(update.y, connector.y);
        });
      }
      const updateMethod = this.renderUpdateMethods[update.type];
      return connectorsToRender.concat(
        updateMethod(update.y, update.popup, update.content)
      );
    });
  }

  render() {
    const { updates } = this.props;

    return (
      <div>
        <svg width="100%" height="700" viewBox="0 0 400 700">
          <text x={timelines.featureOne + 10} y="20">
            feature
          </text>
          <text
            x={timelines.develop - 25}
            y="20"
            style={{ fontWeight: 'bold' }}
          >
            develop
          </text>
          <text x={timelines.release - 20} y="20">
            release
          </text>
          <text x={timelines.hotfix - 25} y="20">
            hotfixes
          </text>
          <text x={timelines.master - 20} y="20" style={{ fontWeight: 'bold' }}>
            master
          </text>

          {this.renderTimeline(timelines.featureOne, timelines.top)}
          {this.renderTimeline(timelines.featureTwo, timelines.top)}
          {this.renderTimeline(timelines.develop, timelines.top)}
          {this.renderTimeline(timelines.release, timelines.top)}
          {this.renderTimeline(timelines.hotfix, timelines.top)}
          {this.renderTimeline(timelines.master, timelines.top)}

          {updates && this.renderUpdates(updates)}
        </svg>
      </div>
    );
  }
}

export default Flow;
