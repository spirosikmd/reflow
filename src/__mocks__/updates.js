export default [
  {
    type: 'master',
    y: 50,
    from: [],
    popup: true,
    content: 'Tag 0.1',
  },
  {
    type: 'develop',
    y: 100,
    from: [
      {
        type: 'master',
        y: 50,
      },
    ],
  },
  {
    type: 'develop',
    y: 150,
    from: [
      {
        type: 'develop',
        y: 100,
      },
    ],
  },
  {
    type: 'featureOne',
    y: 180,
    popup: true,
    content: 'Major feature for next release',
    from: [
      {
        type: 'develop',
        y: 150,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 190,
    popup: true,
    content: 'Feature for future release',
    from: [
      {
        type: 'develop',
        y: 150,
      },
    ],
  },
  {
    type: 'hotfix',
    y: 170,
    popup: true,
    content: 'Severe bug fixed for production: hotfix 0.2',
    from: [
      {
        type: 'master',
        y: 50,
      },
    ],
  },
  {
    type: 'develop',
    y: 200,
    popup: true,
    content: 'Incorporate bugfix in develop',
    from: [
      {
        type: 'hotfix',
        y: 170,
      },
      {
        type: 'develop',
        y: 150,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 225,
    from: [
      {
        type: 'featureTwo',
        y: 190,
      },
    ],
  },
  {
    type: 'master',
    y: 220,
    popup: true,
    content: 'Tag 0.2',
    from: [
      {
        type: 'master',
        y: 50,
      },
      {
        type: 'hotfix',
        y: 170,
      },
    ],
  },
  {
    type: 'featureOne',
    y: 240,
    from: [
      {
        type: 'featureOne',
        y: 180,
      },
    ],
  },
  {
    type: 'develop',
    y: 250,
    from: [
      {
        type: 'featureTwo',
        y: 225,
      },
      {
        type: 'develop',
        y: 200,
      },
    ],
  },
  {
    type: 'release',
    y: 290,
    popup: true,
    content: 'Start of release branch for 1.0. From this point on next release means the release after 1.0',
    from: [
      {
        type: 'develop',
        y: 250,
      },
    ],
  },
  {
    type: 'release',
    y: 340,
    from: [
      {
        type: 'release',
        y: 290,
      },
    ],
  },
  {
    type: 'featureOne',
    y: 370,
    from: [
      {
        type: 'featureOne',
        y: 240,
      },
    ],
  },
  {
    type: 'develop',
    y: 370,
    from: [
      {
        type: 'develop',
        y: 250,
      },
      {
        type: 'release',
        y: 340,
      },
    ],
  },
  {
    type: 'release',
    y: 395,
    popup: true,
    content: 'Only bugfixes',
    from: [
      {
        type: 'release',
        y: 340,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 430,
    from: [
      {
        type: 'develop',
        y: 370,
      },
    ],
  },
  {
    type: 'release',
    y: 450,
    popup: true,
    content: 'Bugfixes from release branch may be continuously merged back into develop',
    from: [
      {
        type: 'release',
        y: 395,
      },
    ],
  },
  {
    type: 'master',
    y: 470,
    popup: true,
    content: 'Tag 1.0',
    from: [
      {
        type: 'release',
        y: 450,
      },
      {
        type: 'master',
        y: 220,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 480,
    from: [
      {
        type: 'featureTwo',
        y: 430,
      },
    ],
  },
  {
    type: 'develop',
    y: 500,
    from: [
      {
        type: 'release',
        y: 450,
      },
      {
        type: 'develop',
        y: 370,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 520,
    from: [
      {
        type: 'featureTwo',
        y: 480,
      },
    ],
  },
  {
    type: 'featureOne',
    y: 530,
    from: [
      {
        type: 'featureOne',
        y: 370,
      },
    ],
  },
  {
    type: 'develop',
    y: 570,
    from: [
      {
        type: 'develop',
        y: 500,
      },
      {
        type: 'featureOne',
        y: 530,
      },
      {
        type: 'featureTwo',
        y: 520,
      },
    ],
  },
  {
    type: 'release',
    y: 600,
    from: [
      {
        type: 'develop',
        y: 570,
      },
    ],
  },
  {
    type: 'develop',
    y: 625,
    from: [
      {
        type: 'develop',
        y: 570,
      },
      {
        type: 'release',
        y: 600,
      },
    ],
  },
  {
    type: 'master',
    y: 650,
    from: [
      {
        type: 'release',
        y: 600,
      },
      {
        type: 'master',
        y: 470,
      },
    ],
  },
];
