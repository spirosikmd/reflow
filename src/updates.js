export default [
  {
    type: 'master',
    y: 50,
    popup: true,
    content: 'Tag 0.1',
    connectors: [
      {
        type: 'develop',
        y: 100,
      },
      {
        type: 'master',
        y: 220,
      },
    ],
  },
  {
    type: 'develop',
    y: 100,
    connectors: [
      {
        type: 'develop',
        y: 150,
      },
    ],
  },
  {
    type: 'develop',
    y: 150,
    connectors: [
      {
        type: 'develop',
        y: 200,
      },
      {
        type: 'featureTwo',
        y: 190,
      },
      {
        type: 'featureOne',
        y: 180,
      },
    ],
  },
  {
    type: 'featureOne',
    y: 180,
    popup: true,
    content: 'Major feature for next release',
    connectors: [
      {
        type: 'featureOne',
        y: 240,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 190,
    popup: true,
    content: 'Feature for future release',
    connectors: [
      {
        type: 'featureTwo',
        y: 220,
      },
    ],
  },
  {
    type: 'hotfix',
    y: 170,
    popup: true,
    content: 'Severe bug fixed for production: hotfix 0.2',
    connectors: [
      {
        type: 'develop',
        y: 200,
      },
      {
        type: 'master',
        y: 220,
      },
    ],
  },
  {
    type: 'develop',
    y: 200,
    popup: true,
    content: 'Incorporate bugfix in develop',
    connectors: [
      {
        type: 'develop',
        y: 250,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 220,
    connectors: [
      {
        type: 'develop',
        y: 250,
      },
    ],
  },
  {
    type: 'master',
    y: 220,
    popup: true,
    content: 'Tag 0.2',
    connectors: [
      {
        type: 'master',
        y: 470,
      },
    ],
  },
  {
    type: 'featureOne',
    y: 240,
    connectors: [
      {
        type: 'featureOne',
        y: 370,
      },
    ],
  },
  {
    type: 'develop',
    y: 250,
    connectors: [
      {
        type: 'release',
        y: 290,
      },
      {
        type: 'develop',
        y: 370,
      },
    ],
  },
  {
    type: 'release',
    y: 290,
    popup: true,
    content: 'Start of release branch for 1.0. From this point on next release means the release after 1.0',
    connectors: [
      {
        type: 'release',
        y: 340,
      },
    ],
  },
  {
    type: 'release',
    y: 340,
    connectors: [
      {
        type: 'develop',
        y: 370,
      },
      {
        type: 'release',
        y: 410,
      },
    ],
  },
  {
    type: 'featureOne',
    y: 370,
    connectors: [
      {
        type: 'featureOne',
        y: 520,
      },
    ],
  },
  {
    type: 'develop',
    y: 370,
    connectors: [
      {
        type: 'develop',
        y: 500,
      },
      {
        type: 'featureTwo',
        y: 430,
      },
    ],
  },
  {
    type: 'release',
    y: 410,
    popup: true,
    content: 'Only bugfixes',
    connectors: [
      {
        type: 'release',
        y: 450,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 430,
    connectors: [
      {
        type: 'featureTwo',
        y: 480,
      },
    ],
  },
  {
    type: 'release',
    y: 450,
    popup: true,
    content: 'Bugfixes from release branch may be continuously merged back into develop',
    connectors: [
      {
        type: 'master',
        y: 470,
      },
      {
        type: 'develop',
        y: 500,
      },
    ],
  },
  {
    type: 'master',
    y: 470,
    popup: true,
    content: 'Tag 1.0',
    connectors: [
      {
        type: 'master',
        y: 620,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 480,
    connectors: [
      {
        type: 'featureTwo',
        y: 520,
      },
    ],
  },
  {
    type: 'develop',
    y: 500,
    connectors: [
      {
        type: 'develop',
        y: 550,
      },
    ],
  },
  {
    type: 'featureTwo',
    y: 520,
    connectors: [
      {
        type: 'develop',
        y: 550,
      },
    ],
  },
  {
    type: 'featureOne',
    y: 530,
    connectors: [
      {
        type: 'develop',
        y: 550,
      },
    ],
  },
  {
    type: 'develop',
    y: 550,
    connectors: [
      {
        type: 'develop',
        y: 600,
      },
      {
        type: 'release',
        y: 570,
      },
    ],
  },
  {
    type: 'release',
    y: 570,
    connectors: [
      {
        type: 'develop',
        y: 600,
      },
      {
        type: 'master',
        y: 620,
      },
    ],
  },
  {
    type: 'develop',
    y: 600,
  },
  {
    type: 'master',
    y: 620,
  },
];
