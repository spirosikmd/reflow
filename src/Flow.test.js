import React from 'react';
import renderer from 'react-test-renderer';
import Flow from './Flow';
import updates from './__mocks__/updates';

it('App renders', () => {
  const component = renderer.create(<Flow updates={updates} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
