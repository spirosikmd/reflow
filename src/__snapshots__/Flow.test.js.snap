// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<div>
  <svg
    height="700"
    viewBox="0 0 400 700"
    width="100%"
  >
    <defs>
      <marker
        id="arrow"
        markerHeight="10"
        markerUnits="strokeWidth"
        markerWidth="10"
        orient="auto"
        refX="18"
        refY="3"
      >
        <path
          d="M0,0 L0,6 L9,3 z"
          fill="#000"
        />
      </marker>
    </defs>
    <text
      x={60}
      y="20"
    >
      feature
    </text>
    <text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
      x={145}
      y="20"
    >
      develop
    </text>
    <text
      x={210}
      y="20"
    >
      release
    </text>
    <text
      x={265}
      y="20"
    >
      hotfixes
    </text>
    <text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
      x={330}
      y="20"
    >
      master
    </text>
    <line
      style={
        Object {
          "stroke": "rgb(208, 208, 208)",
          "strokeWidth": 1,
        }
      }
      x1={50}
      x2={50}
      y1={30}
      y2={680}
    />
    <line
      style={
        Object {
          "stroke": "rgb(208, 208, 208)",
          "strokeWidth": 1,
        }
      }
      x1={110}
      x2={110}
      y1={30}
      y2={680}
    />
    <line
      style={
        Object {
          "stroke": "rgb(208, 208, 208)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={170}
      y1={30}
      y2={680}
    />
    <line
      style={
        Object {
          "stroke": "rgb(208, 208, 208)",
          "strokeWidth": 1,
        }
      }
      x1={230}
      x2={230}
      y1={30}
      y2={680}
    />
    <line
      style={
        Object {
          "stroke": "rgb(208, 208, 208)",
          "strokeWidth": 1,
        }
      }
      x1={290}
      x2={290}
      y1={30}
      y2={680}
    />
    <line
      style={
        Object {
          "stroke": "rgb(208, 208, 208)",
          "strokeWidth": 1,
        }
      }
      x1={350}
      x2={350}
      y1={30}
      y2={680}
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={350}
      x2={170}
      y1={50}
      y2={100}
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={350}
      x2={290}
      y1={50}
      y2={170}
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={350}
      x2={350}
      y1={50}
      y2={220}
    />
    <circle
      cursor="help"
      cx={350}
      cy={50}
      fill="#39E4F7"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={170}
      y1={100}
      y2={150}
    />
    <circle
      cx={170}
      cy={100}
      fill="#FEE148"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={50}
      y1={150}
      y2={180}
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={110}
      y1={150}
      y2={190}
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={170}
      y1={150}
      y2={200}
    />
    <circle
      cx={170}
      cy={150}
      fill="#FEE148"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={50}
      x2={50}
      y1={180}
      y2={240}
    />
    <circle
      cursor="help"
      cx={50}
      cy={180}
      fill="#FC73BF"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={50}
      x2={50}
      y1={240}
      y2={370}
    />
    <circle
      cx={50}
      cy={240}
      fill="#FC73BF"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={50}
      x2={50}
      y1={370}
      y2={530}
    />
    <circle
      cx={50}
      cy={370}
      fill="#FC73BF"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={50}
      x2={170}
      y1={530}
      y2={570}
    />
    <circle
      cx={50}
      cy={530}
      fill="#FC73BF"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={110}
      x2={110}
      y1={190}
      y2={225}
    />
    <circle
      cursor="help"
      cx={110}
      cy={190}
      fill="#FC73BF"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={110}
      x2={170}
      y1={225}
      y2={250}
    />
    <circle
      cx={110}
      cy={225}
      fill="#FC73BF"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={170}
      y1={200}
      y2={250}
    />
    <circle
      cursor="help"
      cx={170}
      cy={200}
      fill="#FEE148"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={170}
      y1={250}
      y2={370}
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={230}
      y1={250}
      y2={290}
    />
    <circle
      cx={170}
      cy={250}
      fill="#FEE148"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={170}
      y1={370}
      y2={500}
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={110}
      y1={370}
      y2={430}
    />
    <circle
      cx={170}
      cy={370}
      fill="#FEE148"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={170}
      y1={500}
      y2={570}
    />
    <circle
      cx={170}
      cy={500}
      fill="#FEE148"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={170}
      y1={570}
      y2={625}
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={170}
      x2={230}
      y1={570}
      y2={600}
    />
    <circle
      cx={170}
      cy={570}
      fill="#FEE148"
      r="10"
      stroke="black"
    />
    <circle
      cx={170}
      cy={625}
      fill="#FEE148"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={110}
      x2={110}
      y1={430}
      y2={480}
    />
    <circle
      cx={110}
      cy={430}
      fill="#FC73BF"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={110}
      x2={110}
      y1={480}
      y2={520}
    />
    <circle
      cx={110}
      cy={480}
      fill="#FC73BF"
      r="10"
      stroke="black"
    />
    <circle
      cx={110}
      cy={520}
      fill="#FC73BF"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={230}
      x2={230}
      y1={290}
      y2={340}
    />
    <circle
      cursor="help"
      cx={230}
      cy={290}
      fill="#57C132"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={230}
      x2={230}
      y1={340}
      y2={395}
    />
    <circle
      cx={230}
      cy={340}
      fill="#57C132"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={230}
      x2={230}
      y1={395}
      y2={450}
    />
    <circle
      cursor="help"
      cx={230}
      cy={395}
      fill="#57C132"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={230}
      x2={350}
      y1={450}
      y2={470}
    />
    <circle
      cursor="help"
      cx={230}
      cy={450}
      fill="#57C132"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={230}
      x2={350}
      y1={600}
      y2={650}
    />
    <circle
      cx={230}
      cy={600}
      fill="#57C132"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={350}
      x2={350}
      y1={470}
      y2={650}
    />
    <circle
      cursor="help"
      cx={350}
      cy={470}
      fill="#39E4F7"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
    <circle
      cx={350}
      cy={650}
      fill="#39E4F7"
      r="10"
      stroke="black"
    />
    <line
      markerEnd="url(#arrow)"
      style={
        Object {
          "stroke": "rgb(0, 0, 0)",
          "strokeWidth": 1,
        }
      }
      x1={290}
      x2={350}
      y1={170}
      y2={220}
    />
    <circle
      cursor="help"
      cx={290}
      cy={170}
      fill="#FB5B68"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
    <circle
      cursor="help"
      cx={350}
      cy={220}
      fill="#39E4F7"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      r="10"
      stroke="black"
      strokeWidth="3"
    />
  </svg>
</div>
`;
